### PR TITLE
fix(workspace): dendron can hang when trying to provide hover for large non-dendron file

### PR DESCRIPTION
**Pull request by kevinslin**
fix(workspace): dendron can hang when trying to provide hover for large non-dendron file

This currently happens because `ReferenceHoverProvider` can be triggered for all files. This function calls `getReferenceAtPosition` which will attempt to parse the file using `unified` to find hover positions. This will blow up spectacularly when parsing large files (eg. Dendron logs) and cause the vscode process to hang.

This fix adds a check in `getReferenceAtPosition` to not parse the file if it is not a dendron note or in a dendron workspace. It also provides this check across most of the symbol provider methods in Dendron. List of changes below:

Changes:
- convert `getReferenceAtPosition` to async, do not run for non-dendron note
- ReferenceHoverProvider restrict to dendron note (cause of hanging when hovering over log file)
- ReferenceProvider restrict to dendron note (was causing silent exception)
- updated signatures in the following files which called `getReferenceAtPosition`
  - [[../packages/plugin-core/src/commands/ConvertLink.ts]]
  - [[../packages/plugin-core/src/commands/GotoNote.ts]]
  - [[../packages/plugin-core/src/features/DefinitionProvider.ts]]
  - [[../packages/plugin-core/src/features/RenameProvider.ts]]

New Utilities:
- `Awaited` utility, unwraps `promise` type from type signature
    - loc: packages/common-all/src/types/typesv2.ts
- `extractFMTags`
    - loc: packages/engine-server/src/markdown/remark/utils.ts
- `isActiveAndIsDendronNote`
    - loc: packages/plugin-core/src/ExtensionProvider.ts

***
- before: 47
- after: 49

This change introduces the following circular dependencies. This is caused by:
- [[../packages/plugin-core/src/dendronExtensionInterface.ts]] > [[../packages/plugin-core/src/WorkspaceWatcher.ts]]

To move forward, we should create a `WorkspaceWatcher` interface. Would move this as a next item since this is a high severity fix

```
9) src/ExtensionProvider.ts > src/dendronExtensionInterface.ts > src/WorkspaceWatcher.ts > src/fileWatcher.ts > src/workspace.ts > src/commandFactory.ts > src/commands/GotoNote.ts

23) src/ExtensionProvider.ts > src/dendronExtensionInterface.ts > src/WorkspaceWatcher.ts > src/fileWatcher.ts > src/workspace.ts > src/commandFactory.ts > src/commands/GotoNote.ts > src/utils/md.ts
```

***

# Dendron Extended PR Checklist

### Basics

- [x] code should follow [Code Conventions](dev.process.code)
- [.] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](dev.process.code.best-practices).
- [x] sticking to existing conventions instead of creating new ones 
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended

- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed 

## Instrumentation
NA

### Basics

- [~] [Write Tests](dev.process.qa.test) 
- [.] [Confirm existing tests pass](dev.process.qa.test)
- [x] [Confirm manual testing](dev.process.qa.test) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [~] If your tests changes an existing snapshot, snapshots have been [updated](dev.process.qa.test)

### Extended

NA

### Basics

- [~] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR

## 

### Basics

- [.] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](dev) or [Packages](pkg)

add new utilities, will be added to docs once this pr is approved

## 

### Extended

- [ ]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](dev.process.close-loop)
  - eg. [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

### CODE REVIEW

**Comment by jonathanyeung** [comment]
Re: circular deps - i think it's fine.  Already tracking the workspace watcher circ dep via [[Circ Dep Workspace Watcher|dendron://private/task.dev.2022.01.11.circ-dep-workspace-watcher]].  These should be a dupe to the existing cycles.

**Review by jonathanyeung — CHANGES_REQUESTED** [review]

File: packages/plugin-core/src/test/suite-integ/ReferenceHoverProvider.test.ts [hunk 1]
```diff
@@ -0,0 +1,1252 @@
...
+async function provideForNonNote(editor: vscode.TextEditor) {
+  return provide({ editor, pos: new vscode.Position(0, 2) });
+}
+
+suite("GIVEN ReferenceHoverProvider", function () {
+  const ctx = setupBeforeAfter(this);
+
+  describe("AND GIVEN non-dendron file", () => {
+    describeMultiWS(
+      "AND WHEN try to find references",
```
Comment (SeriousBug) [inline packages/plugin-core/src/test/suite-integ/ReferenceHoverProvider.test.ts:50]: nitpick: I think the messages here could be clearer, like "WHEN used in a non-dendron file", "AND the file is empty", "THEN doesn't provide for this file", so that it reads like a sentence when put together.

File: packages/plugin-core/src/utils/md.ts [hunk 2]
```diff
@@ -237,11 +235,22 @@ export type getReferenceAtPositionResp = {
   refText: string;
 };
 
-export const getReferenceAtPosition = (
-  document: vscode.TextDocument,
-  position: vscode.Position,
-  opts?: { partial?: boolean; allowInCodeBlocks: boolean }
-): getReferenceAtPositionResp | null => {
+export const getReferenceAtPosition = async ({
```
Comment (SeriousBug) [inline packages/plugin-core/src/utils/md.ts:238]: could we use a named function here like `export async function getReferenceAtPosition` so the name will show up in backtraces?

File: packages/plugin-core/src/utils/md.ts [hunk 3]
```diff
@@ -313,18 +325,19 @@ export const getReferenceAtPosition = (
...
+    if (
+      !(await WorkspaceUtils.isDendronNote({
+        wsRoot,
+        vaults,
+        fpath: document.uri.fsPath,
+      }))
+    ) {
+      return null;
+    }
+    const maybeTags = RemarkUtils.extractFMTags(document.getText());
```
Comment (SeriousBug) [inline packages/plugin-core/src/utils/md.ts:338]: No need to add block this PR for it, but we could also limit how much of the text we grab here, like `document.getText(new vscode.Range(0, 0, 300, 0))`. This could cause confusion if someone does legitimately write a note that has more than 300 lines of frontmatter, but it would also help in actual notes that are very long.

File: packages/plugin-core/src/test/suite-integ/ConvertLink.test.ts [hunk 4]
```diff
@@ -18,6 +19,23 @@ import { WSUtils } from "../../WSUtils";
...
+  position: vscode.Position;
+}) => {
+  const { wsRoot, vaults } = ExtensionProvider.getDWorkspace();
+  const out = await getReferenceAtPosition({
+    document: editor.document,
+    position,
+    wsRoot,
+    vaults,
+  });
+  return out as getReferenceAtPositionResp;
```
Comment (SeriousBug) [inline packages/plugin-core/src/test/suite-integ/ConvertLink.test.ts:36]: It might be good to add a `expect(out).toBeTruthy()` here so the test will fail early if there's an issue.

**Review by SeriousBug — COMMENTED** [review]
Left some comments. The fix looks good to me.

File: packages/plugin-core/src/utils/md.ts (hunk 2 above)
Comment (kevinslin) [reply packages/plugin-core/src/utils/md.ts:238]: updated. and good point. can you write an async about this and update [[Best Practices|dendron://dendron.docs/dev.process.code.best-practices]]?

File: packages/plugin-core/src/utils/md.ts (hunk 3 above)
Comment (kevinslin) [reply packages/plugin-core/src/utils/md.ts:338]: in terms of parsing text, vscode and a lot of editors have a hard limit of what they're willing to parse (eg. don't pass if file is larger than 1MB). we should do something similar and have a metric for how many times we hit that. will add as a next item

File: packages/plugin-core/src/test/suite-integ/ConvertLink.test.ts (hunk 4 above)
Comment (kevinslin) [reply packages/plugin-core/src/test/suite-integ/ConvertLink.test.ts:36]: added a check

File: packages/plugin-core/src/workspace.ts [hunk 5]
```diff
@@ -304,6 +304,10 @@ export class DendronExtension implements IDendronExtension {
     return false;
   }
 
+  isActiveAndIsDendronNote(_fpath: string): boolean {
+    throw new Error("not implemented");
```
Comment (jonathanyeung) [inline packages/plugin-core/src/workspace.ts:308]: So to follow through - let's move the logic from the current impl in `ExtensionProvider` to here, and then replace the impl in ExtensionProvider with 

```typescript
  static isActiveAndIsDendronNote(fpath: string) {
    return ExtensionProvider.getExtension().isActiveAndIsDendronNote(fPath);
  }
```

**Review by jonathanyeung — CHANGES_REQUESTED** [review]

**Review by jonathanyeung — APPROVED** [review]
